### PR TITLE
Feature/take upgrade cost into account

### PR DESCRIPTION
**Pull request by Urky**
When merged this pull request will: 

- Fixed the bug upgrading past the disciplinarian perk
- Fixed a bug where equipmentproperties were not taken into account
- Added Split upgradepaths functionality
- Added Playerdecision switch to disable all logic behind multipath upgrading

### CODE REVIEW

File: Extension/Global/Strings.cs [hunk 1]
```diff
@@ -24,6 +24,12 @@ public static class Strings
         public const string SETTINGS_PREFER_SHIELD_TOOLTIP = "Prefer infantry with shields when upgrading to infantry troop classes.";
         public const string SETTINGS_RANGED_PREFERENCE_NAME = "Prefer ranged with specific weapons";
         public const string SETTINGS_RANGED_PREFERENCE_TOOLTIP = "Prefer specific ranged weapons when upgrading to ranged troop classes. (0 for no preference, 1 for Bows, 2 for crossbows)";
+        public const string SETTINGS_BUFFER_DAILY_COST_NAME = "Buffer days to pay daily costs";
+        public const string SETTINGS_BUFFER_DAILY_COST_TOOLTIP = "Number of days to pay daily costs upgrading should not use.";
```
Comment (MLNW) [inline Extension/Global/Strings.cs]: This has to be reworded

File: Extension/Global/Strings.cs [hunk 2]
```diff
@@ -24,6 +24,12 @@ public static class Strings
...
+        public const string SETTINGS_SPLIT_UPGRADES_NAME = "Split upgrades if decision can't be made";
+        public const string SETTINGS_SPLIT_UPGRADES_TOOLTIP = "Splits upgradepaths when no decision can be made. NOTE: decisions will be made based on culture strengths. More info on the mod's page.";
```
Comment (MLNW) [inline Extension/Global/Strings.cs]: Sounds like this is the toggle for the whole feature?

File: Extension/MultipathUpgrade/Model/AvailableResources.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+using TaleWorlds.Core;
+
+namespace YAPO.MultipathUpgrade.Model
+{
+    public class AvailableResources
+    {
+        public AvailableResources(bool hotkeyControl)
+        {
+            int cost = 0;
+            if (!hotkeyControl)
```
Comment (MLNW) [inline Extension/MultipathUpgrade/Model/AvailableResources.cs]:
```suggestion
        public AvailableResources()
        {
            int cost = 0;
            if (!States.HotkeyControl)
```

File: Extension/MultipathUpgrade/Model/AvailableResources.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                int dailyCost = GetMostAccurateDailyCost();
+                cost = dailyCost * YapoSettings.Instance.DaysToPayDailyCostsBuffer;
+            }
+
+            AvailableGold = Hero.MainHero.Clan.Gold + cost;
```
Comment (MLNW) [inline Extension/MultipathUpgrade/Model/AvailableResources.cs:22]: Shouldn't the cost be subtracted from the available gold?
```suggestion
            AvailableGold = Hero.MainHero.Clan.Gold - cost;
```

File: Extension/MultipathUpgrade/Model/AvailableResources.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        public int AvailableGold { get; set; }
+        public Dictionary<ItemCategory, int> ItemsOfCategoryWithCount { get; set; } = new Dictionary<ItemCategory, int>();
```
Comment (MLNW) [inline Extension/MultipathUpgrade/Model/AvailableResources.cs]: I usually put members like these at the top of a class. Not sure how @tbeswick96 does it.

Comment (MLNW) [inline Extension/MultipathUpgrade/Model/AvailableResources.cs]:
```suggestion
        public int AvailableGold { get; private set; }
        public Dictionary<ItemCategory, int> ItemsOfCategoryWithCount { get; } = new Dictionary<ItemCategory, int>();
```

File: Extension/MultipathUpgrade/Model/AvailableResources.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            return 0;
+        }
+
+        public void UpdateAvailableResources(PartyCharacterVM member, int troopsToUpgrade)
+        {
+            AvailableGold -= member.Character.UpgradeCost(PartyBase.MainParty, 0) * troopsToUpgrade;
+
+            if (member.Character.UpgradeRequiresItemFromCategory != null)
+            {
+                ItemsOfCategoryWithCount[member.Character.UpgradeRequiresItemFromCategory] -= troopsToUpgrade;
```
Comment (MLNW) [inline Extension/MultipathUpgrade/Model/AvailableResources.cs]: What if there aren't enough items of this type to upgrade `troopsToUpgrade`?

**Comment by MLNW** [comment]
Closes #5 & #10

File: Extension/Global/Strings.cs (hunk 2 above)
Comment (Urky) [reply Extension/Global/Strings.cs]: Yeah, which reminds me: @tbeswick96 Could you include Information regarding the culturestrengths on the mod's page? Ill type it myself dont worry (or just copy from my mod)

File: Extension/MultipathUpgrade/Model/AvailableResources.cs (hunk 4 above)
Comment (Urky) [reply Extension/MultipathUpgrade/Model/AvailableResources.cs:22]: Yeah, but the cost is -x and y - -x ads it up

File: Extension/MultipathUpgrade/Model/AvailableResources.cs (hunk 5 above)
Comment (Urky) [reply Extension/MultipathUpgrade/Model/AvailableResources.cs]: Properties go below the ctor

File: Extension/MultipathUpgrade/Model/AvailableResources.cs (hunk 6 above)
Comment (Urky) [reply Extension/MultipathUpgrade/Model/AvailableResources.cs]: This is an update after the fact. However beforehand the resourcecost is determined in TroopActionService.CalculateCountOfTroopsThatCanBeUpgraded. That's where the itemcost is decided. This is to update that amount

File: Extension/Global/Strings.cs (hunk 1 above)
Comment (Urky) [reply Extension/Global/Strings.cs]: Keep enough gold for X days of troop wages? (thanks Autotrader)

File: Extension/YetAnotherPartyOrganiser.csproj [hunk 7]
```diff
@@ -1,357 +1,357 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (MLNW) [inline Extension/YetAnotherPartyOrganiser.csproj]: What changed here?

Comment (tbeswick96) [reply Extension/YetAnotherPartyOrganiser.csproj]: Think rider and VS save csproj files with different indentation. This'll keep happening unless we can define the indent width in an editorconfig file which works with VS and Rider natively

File: Extension/Global/Strings.cs [hunk 8]
```diff
@@ -24,6 +24,12 @@ public static class Strings
         public const string SETTINGS_PREFER_SHIELD_TOOLTIP = "Prefer infantry with shields when upgrading to infantry troop classes.";
         public const string SETTINGS_RANGED_PREFERENCE_NAME = "Prefer ranged with specific weapons";
         public const string SETTINGS_RANGED_PREFERENCE_TOOLTIP = "Prefer specific ranged weapons when upgrading to ranged troop classes. (0 for no preference, 1 for Bows, 2 for crossbows)";
+        public const string SETTINGS_BUFFER_DAILY_COST_NAME = "Keep enough gold for X days of troop wages?";
+        public const string SETTINGS_BUFFER_DAILY_COST_TOOLTIP = "Reserves gold to pay daily costs for X amount of days.";
+        public const string SETTINGS_SPLIT_UPGRADES_NAME = "Split upgrades if decision can't be made";
+        public const string SETTINGS_SPLIT_UPGRADES_TOOLTIP = "Splits upgradepaths when no decision can be made. NOTE: decisions will be made based on culture strengths.";
+        public const string SETTINGS_PLAYER_DECISION_NAME = "Lets player decide on every upgrade split path";
+        public const string SETTINGS_PLAYER_DECISION_TOOLTIP = "Lets player decide on every upgrade when two upgrade options are available";
```
Comment (MLNW) [inline Extension/Global/Strings.cs]: Still not exactly sure what this is supposed to enable/disable/decide

**Comment by MLNW** [comment]
@tbeswick96 might be best if you get around to give a review

File: Extension/Global/Strings.cs (hunk 8 above)
Comment (Urky) [reply Extension/Global/Strings.cs]: All the split upgrade paths are now decided by an algorithm. That would be disabled.

File: Extension/Global/Strings.cs [hunk 9]
```diff
@@ -24,6 +24,12 @@ public static class Strings
         public const string SETTINGS_PREFER_SHIELD_TOOLTIP = "Prefer infantry with shields when upgrading to infantry troop classes.";
         public const string SETTINGS_RANGED_PREFERENCE_NAME = "Prefer ranged with specific weapons";
         public const string SETTINGS_RANGED_PREFERENCE_TOOLTIP = "Prefer specific ranged weapons when upgrading to ranged troop classes. (0 for no preference, 1 for Bows, 2 for crossbows)";
+        public const string SETTINGS_BUFFER_DAILY_COST_NAME = "Keep enough gold for X days of troop wages?";
+        public const string SETTINGS_BUFFER_DAILY_COST_TOOLTIP = "Reserves gold to pay daily costs for X amount of days.";
+        public const string SETTINGS_SPLIT_UPGRADES_NAME = "Split upgrades if decision can't be made";
+        public const string SETTINGS_SPLIT_UPGRADES_TOOLTIP = "Splits upgradepaths when no decision can be made. NOTE: decisions will be made based on culture strengths.";
+        public const string SETTINGS_PLAYER_DECISION_NAME = "Lets player decide on every upgrade split path";
```
Comment (tbeswick96) [inline Extension/Global/Strings.cs]: If I've interpreted this right, this disables multipath decision making correct? So behaviour with this enabled is single path troops are upgraded on upgrade all, but multipath are skipped?
And disabled all upgrades are resolved based on other settings?

Comment (Urky) [reply Extension/Global/Strings.cs]: Playerdecision will only do the upgrades that do not branch, where the player will need to make a decision

**Review by tbeswick96 — APPROVED** [review]

**Pull request merged**